tree/dataframe/test/dataframe_regression_tests.hxx

```diff
@@ -146,3 +146,59 @@ TEST(TEST_CATEGORY, ReadWriteTVector3)
    gSystem->Unlink(snap_fname.c_str());
    gSystem->Unlink(filename.c_str());
 }
+
+TEST(TEST_CATEGORY, PolymorphicTBranchObject)
+{
+   const std::string filename = "polymorphictbranchobject.root";
+   {
+      TFile f(filename.c_str(), "recreate");
+      TTree t("t", "t");
+      TObject *o = nullptr;
+      t.Branch("o", &o, 32000, 0); // must be unsplit to generate a TBranchObject
+
+      // Fill branch with different concrete types
+      TNamed name("name", "title");
+      TList list;
+      list.Add(&name);
+      o = &list;
+      t.Fill();
+      TH1D h("h", "h", 100, 0, 100);
+      h.Fill(42);
+      o = &h;
+      t.Fill();
+      o = nullptr;
+
+      t.Write();
+   }
+
+   auto checkEntries = [](const TObject &obj, ULong64_t entry) {
+      if (entry == 0) {
+         EXPECT_STREQ(obj.ClassName(), "TList");
+         auto &asList = static_cast<const TList &>(obj);
```

Comment:
```suggestion
         auto &asList = dynamic_cast<const TList &>(obj);
```
to catch errors? If the cast fails, something is seriously broken, so maybe good to test.

Reply:
good idea, will switch to dynamic casts and add checks

Reply:
No need to check. dynamic throws when used on reference.

Reply:
my lack of experience with dynamic casts is showing xD